Add SeparatorMenu class for menu separators and enhance rendering logic

abdelaziz-mahdy opened the pull request:
Introduce a new SeparatorMenu class to render horizontal line separators in menus. Update the menu system to integrate SeparatorMenu and improve custom row view handling in NestedTreeView and TreeView.

abdelaziz-mahdy committed:
Add SeparatorMenu class and integrate into menu system

- Introduced a new SeparatorMenu class for rendering horizontal line separators in menus.
- Updated menu definitions to include SeparatorMenu.
- Modified NestedTreeView and TreeView to handle SeparatorMenu rendering appropriately.
- Enhanced menu item rendering logic to accommodate custom row views for handlers.

abdelaziz-mahdy committed:
Enhance SeparatorMenu integration and custom row view handling

- Added createRowView method to SeparatorMenu for improved rendering.
- Updated NestedTreeView and TreeView to utilize createRowView for custom row views.
- Removed legacy code related to direct SeparatorMenu handling in view rendering.

sarthak-marwaha commented on src/foam/core/menu/Menu.js:
```diff
@@ -75,7 +75,9 @@
           [ 'foam.core.menu.TabsMenu',         'Tabs' ],
-          [ 'foam.core.menu.ViewMenu',         'View' ]
+          [ 'foam.core.menu.ViewMenu',         'View' ],
+          [ 'foam.core.menu.SeparatorMenu',    'Separator' ],
```
Trailing comma

sarthak-marwaha commented:
@abdelaziz-mahdy Dont really like the idea that we need to have a menu that is a separator, would crowd up journals and there would be no real config for this menu. Can you explain the use case? Maybe provide some screenshots of where you want this to happen and we can find a solution, or workshop this one.

sarthak-marwaha commented on src/foam/core/menu/SeparatorMenu.js:
```diff
@@ -0,0 +1,42 @@
+        ^ {
+          height: 1px;
+          width: 100%;
+          background: #e0e0e0;
```
Should use cssTokens

sarthak-marwaha commented:
Initial idea would be that menus already have parents, you could check if a menu has a SubMenu handler type (Which is only used by menus that have children and dont launch anything themselves, so esentially acting as grouping menus) then show separators after the submenu's children are displayed. Assuming that works for your usecase.

abdelaziz-mahdy committed:
Enhance SeparatorMenu styling with CSS tokens

- Introduced CSS tokens for separator color, margin, height, and width to improve customization.
- Updated the CSS rules in SeparatorView to utilize the new tokens for better maintainability.

abdelaziz-mahdy committed:
Fix formatting in Menu.js by removing trailing whitespace in SeparatorMenu definition

abdelaziz-mahdy commented:
The ability to set the seprator anywhere using the order will make it more customizable, and for the sub menu that limited in that case i think? since what if we want to seperate the main menus?.

also this is how it looks currently 
<img width="249" alt="image" src="https://github.com/user-attachments/assets/5f771d1e-11af-4b38-96e0-a209b2638535" />

sarthak-marwaha commented:
Yup, thats right. Parents wont work in this case. Let me give it a think, I'll get back to you tomorrow. The problem with this approach is you will end up with 10s of menus called "menu-separator-1","menu-separator-2","menu-separator-3" with no real reason for them to exist. Maybe adding a property to the menu called "group" would be better, would work the same as order but every group gets a separator. Can also expand the group functionality to add headers or extra things like that.

Just thoughts, If you have any ideas to side step menus let me know. Otherwise if we cant come up with something better we can go with this for now

kgrgreer commented on src/foam/u2/view/NestedTreeView.js:
```diff
@@ -54,7 +54,19 @@ foam.CLASS({
       if ( this.translationService ) {
         labelString = self.translationService.getTranslation(foam.locale, self.data.label, self.data.label);
       }
+
```
Why does this have to be here? NestedTreeView isn't Menu specific and we don't want to have to make it tightly coupled to menus or even a specific menu handler. The idea is that we should be able to make menu handlers extensible.

abdelaziz-mahdy replied:
due to its forced styling i couldnt do it with any other way, check : https://github.com/kgrgreer/foam3/pull/3930 where i used row config to handle it, it shows the seprator, but there is forced stlying on it which is not logical, due to its being clickable and had uneeded paddings.

sarthak-marwaha replied:
The more i think about it, the more menu groups sounds like a better idea. They can be a DAO and be journaled, the treeView can do a groupBy on the group property (which wont be Menu specific) and then have styling for when it encounters new groups.

kgrgreer commented on src/foam/u2/view/TreeView.js:
```diff
@@ -243,6 +242,18 @@ foam.CLASS({
         labelString = self.translationService.getTranslation(foam.locale, self.data.label, self.data.label);
       }
 
+      // Check if handler has a custom row view
```
Same comment.

abdelaziz-mahdy committed:
Merge development branch

abdelaziz-mahdy committed:
Implement createRowView method in Menu and update TreeView/NestedTreeView to utilize it for custom row rendering

abdelaziz-mahdy commented:
now it shouldn't be tightly coupled with menu. any object that implements `createRowView` can override it.

kgrgreer commented on src/foam/u2/view/NestedTreeView.js:
```diff
@@ -57,6 +57,18 @@ foam.CLASS({
       }
 
+      // Check if handler has a custom row view
+      if (this.data.createRowView) {
```
See the FOAM style guide:
Should be:
if ( this.data.createdRowView ) {

abdelaziz-mahdy replied:
will try to get a formatter working to avoid this problem in the new prs, if i got it will document it.

kgrgreer commented on src/foam/u2/view/NestedTreeView.js:
```diff
@@ -25,24 +26,23 @@ foam.CLASS({
-          );
+          this.addClass(self.myClass('select-level'))
+            .start(self.Image, { glyph: 'next' })
+            .addClass(self.myClass('toggle-icon'))
```
While you're here, indent this line and the next two spaces.

abdelaziz-mahdy committed:
Refactor NestedTreeView and TreeView for improved readability by adjusting whitespace in conditional statements and method calls.
